File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> computers;

vector<bool> searchConnectedComputers(vector<bool> visited, int start) {
    queue<int> q;
    q.push(start);
    visited[start] = true;
    
    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        
        for (int i = 0; i < computers[curr].size(); i++) {
            if (!visited[i] && computers[curr][i] == 1) {
                visited[i] = true;
                q.push(i);
            }
        }
    }
    
    return visited;
}

int solution(int n, vector<vector<int>> computers) {
    
    ::computers = computers;
    vector<bool> visited(n, false);
    int answer = 0;
    
    for (int i = 0; i < n; i++) {
        
        if (!visited[i]) {
            visited = searchConnectedComputers(visited, i);
            answer++;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    int[][] computers;
    
    public int solution(int n, int[][] computers) {
        
        this.computers = computers;
        int answer = 0;
        boolean[] visited = new boolean[computers.length];
        
        for (int i = 0; i < computers.length; i++) {
            
            if (!visited[i]) {
                visited = searchConnectedComputers(visited, i);
                answer++;
            }
        }
        
        return answer;
    }
    
    private boolean[] searchConnectedComputers(boolean[] visited, int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        visited[start] = true;
        
        while (!q.isEmpty()) {
            int curr = q.poll();
            
            for (int i = 0; i < computers[curr].length; i++) {
                
                if (!visited[i] && computers[curr][i] == 1) {
                    visited[i] = true;
                    q.add(i);
                }
            }
        }
        
        return visited;
    }
}